### ISSUE TITLE
What should be `bin`, and what should be `lib`?

### ISSUE THREAD

**Issue by SimonJF** [links-lang/links#629]
Tom Davey and I are working on a notebook interface (see Jupyter notebooks, for example) for Links. The current design is to write a wrapper application which uses Links as a library, and we can get quite far using `Links_core`.

However, right now, a fair chunk of functionality that would be quite useful is in `bin`, in particular `driver.ml`. Duplicating this code is fine, but a pity.

I wonder whether a better design might be to refactor somewhat in order to have nice things like Prelude loading callable as a library, rather than hardwired into the executable.

**Comment by jstolarek** [comment]
> I wonder whether a better design might be to refactor somewhat in order to have nice things like Prelude loading callable as a library, rather than hardwired into the executable.

That sounds like a right thing to do.

Aside: I always find `bin` name confusing since typically this is a directory where build artefacts are placed. `src` would be a better name IMO.

**Comment by SimonJF** [comment]
> Aside: I always find bin name confusing since typically this is a directory where build artefacts are placed. src would be a better name IMO.

I agree, this confused me in exactly the same way. I think it's a `dune` convention, however.

**Comment by dhil** [comment]
> I wonder whether a better design might be to refactor somewhat in order to have nice things like Prelude loading callable as a library, rather than hardwired into the executable.

For #603 I have implemented a new loader, which contains a bootstrapping phase that loads the Prelude and the built-in `lib.ml` units. Ultimately the loader returns a compilation environment containing the loaded/processed compilation units. I was planning on keeping the wiring in `driver.ml`, but I can move that into `core` instead if need be.

Another thing is, that `core` would really be named `lib`, and our current lib should be named `runtime`.

**Comment by SimonJF** [comment]
@dhil: 

Okay, that seems sensible; a lot of this does seem tied up with the new module system.

Some of the wiring does make sense to be executable-specific -- but I think it needs to be the absolute minimum that is specific to the Links executable. Looking at `driver.ml`, it would seem that a fair bit (if not all of it, in fact) could be moved to `core`. I think it's really lovely to be able to use the compiler as a library and this is a huge advantage of your previous work porting the system to `dune`. What's missing right now is the high-level glue allowing someone to take a raw string and get back a value and some updated environments.

Some of this will depend on how `prelude` is handled in the new module system. I'm imagining your aim is to have it as just a plain old module, as opposed to handled using the current specialised logic?

Since this is tied to the module system I'm inclined to leave this issue as solvable when the module system lands rather than doing anything too huge on it now.

**Comment by dhil** [comment]
@simonjf maybe we can chat about this in person, just make sure I understand your requirements.

**Comment by jamescheney** [comment]
See also: #415.  Reorganizing so that the front-end is reusable for the links notebook server may align with the "reorganize file processing" part.
